### PR TITLE
Replace old datasync walkthrough with the latest one

### PR DESCRIPTION
**Pull request by wtrocki**
## Motivation

Replace the old walkthrough that was covering the only deployment of the data sync and demoing the client-side features. 

A new walkthrough focused on showcasing instant integration and model driven development along with the datasync features.
We are hoping to improve it over the coming months but also we want to replace the old one as it is basing on the angular framework and we currently are supporting the react framework. 
There is no easy way for us to backport the changes so this way we will make sure that we support the latest AMQ Online and Keycloak 

## Notes

This walkthrough is larger than previous and it has multiple steps.
It involves codeready as this was requested to be used with the demos and workshops for mobile/datasync.
Our team was working on a new walkthrough that we have been demoing already and showing the early versions. I'm pushing this early to be able to provide reliable updates that will make things work with the latest SSO and hoping for small incremental updates over the coming weeks

### CODE REVIEW

**Comment by wtrocki** [comment]
CC @tiffanynolan is it possible to test this on some cluster you have?

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+// URLs
+:openshift-console-url: {openshift-host}/console
+:sso-realm-url: {user-sso-url}/auth/admin/solution-patterns/console/index.html
+:data-sync-documentation-url: https://access.redhat.com/documentation/en-us/red_hat_managed_integration/{rhmi-version}/html-single/developing_a_data_sync_app/index
+
+//attributes
+:integreatly-name: Managed Integration
+:data-sync-name: Data Sync
+:data-sync-starter: Data Sync Starter
+:customer-sso-name: SSO
```
Comment (craicoverflow) [inline walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc]: Red Hat SSO?

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+:sectnums:
+
+[time=15]
+== Creating your DataSync project using DataSync Starter template
+
+{data-sync-name} allows you to focus on your business model by giving you the ability
+to generate a fully functional GraphQL based API and Node.js Server and Client Side components.
+
+Developers can use GraphQL types to define the models their application deals with
+and generate the underlying backend that works out of the box with the RHMI services like SSO or AMQ Online.
```
Comment (craicoverflow) [inline walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc]:
```suggestion
and generate the underlying backend that works out of the box with the RHMI services like Red Hat-SSO or AMQ Online.
```

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+
+Using GraphQL subscriptions, your application can receive live updates thanks to GraphQL and the DataSync client allows your application to operate independently of the network connection. 
+
+DataSync-Starter is an application template that gives developers
```
Comment (craicoverflow) [inline walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc]:
```suggestion
DataSync-Starter is an application template that gives developers an
```

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+
+Developers can use GraphQL types to define the models their application deals with
+and generate the underlying backend that works out of the box with the RHMI services like SSO or AMQ Online.
+
+Using GraphQL subscriptions, your application can receive live updates thanks to GraphQL and the DataSync client allows your application to operate independently of the network connection. 
+
+DataSync-Starter is an application template that gives developers 
+opiniated implementation for React Ionic PWA and Mobile application backend by Node.js Server.
+
+DataSync as framework consist of the two major components:
```
Comment (craicoverflow) [inline walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc]: Should this be `DataSync Framework`, or do you mean `DataSync as a framework`?
```suggestion
DataSync Framework consist of the two major components:
```

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+
+* DataSync client (upstream https://offix.dev) 
++
+Can be added to any web or mobile application
+to provide DataSynchronization capabilities
```
Comment (craicoverflow) [inline walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc]:
```suggestion
to provide Data synchronization capabilities
```

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+
+* Node.js Server with code generation capabilties (upstream https://graphback.dev)
```
Comment (craicoverflow) [inline walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc]: No longer code generation capabilities (unless you are talking about client code)

```suggestion
* Node.js Server with out of the box GraphQL API that is backed by MongoDB (upstream https://graphback.dev) 
```

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+DataSync as framework consist of the two major components:
+
+* DataSync client (upstream https://offix.dev) 
++
+Can be added to any web or mobile application
+to provide DataSynchronization capabilities
+
+* Node.js Server with code generation capabilties (upstream https://graphback.dev) 
++
+Server is providing ability to dynamically generate Node.js GraphQL server with DataSynchronization and Conflict resolution capabilties
```
Comment (craicoverflow) [inline walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc]: Not one word:

```suggestion
Server is providing ability to dynamically generate Node.js GraphQL server with Data synchronization and Conflict resolution capabilities 
```

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
++
+The commands are listed in the `MyWorkspace` view available by selecting the box icon in the right top corner of the screen.
+
+=== Creating your DataSync model
+
+Based on a user defined **Model** DataSync generates the underlying GraphQL API and database resources.
+The starter template uses MongoDB as the default datasource for storing application data.
+
+. In your project explorer view on the left,  go to the `./model/task.graphql` file. 
+This file contains an link:https://graphql.org/learn/schema/#object-types-and-fields[GraphQL Schema, window="_blank"]. Here you can define types of the data your application will work with. Graphback will use your model and generate underlying data access methods in the form of GraphQL Queries and Mutations
```
Comment (craicoverflow) [inline walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc]:
```suggestion
This file contains a link:https://graphql.org/learn/schema/#object-types-and-fields[GraphQL Schema, window="_blank"]. Here you can define types of the data your application will work with. Graphback will use your model and generate underlying data access methods in the form of GraphQL Queries and Mutations
```

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc [hunk 9]
```diff
@@ -0,0 +1,435 @@
...
+. Please add a `@datasync` annotation under the `@model` annotation on the `Task` type.
```
Comment (craicoverflow) [inline walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc:101]: I believe this is different now, `@delta`?

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+type Task {
+  id: ID!
+  title: String!
+  description: String!
+  status: TaskStatus
+  address: String
+}
+----
+
+=== Generating your DataSync Node.JS server and React App
```
Comment (craicoverflow) [inline walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc]:
```suggestion
=== Generating your DataSync Node.js server and React App
```

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+** determines what types of CRUD operations will be generated into GraphQL Queries and Mutations in your application.
+** enables and configures various plugins that are used during the code generation process.
+. Make sure that all fields in the `crud` section are enabled 
+. Execute graphback cli command to generate source code:
+`yarn graphback generate`. You can also execute it as predefined `generate source code` command in CodeReady
+. Review `./server/src/schema/schema.qraphql`. 
+This file has the original model and it also contains generated Queries and Mutations. The types of queries and mutations included are based on the `crud` fields in `.graphqlrc.yml`.
+. Review the generated resolver files in `./server/src/resolvers/resolvers.ts`
+This file contains methods used to fetch and modify data. Each individual method uses a 
+preconfigured `MongoDataProvider` which is an abstraction over a MongoDB client. Developers can point resolvers to any datasource.
```
Comment (craicoverflow) [inline walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc]: MongoDBDataProvider
```suggestion
preconfigured `MongoDBDataProvider` which is an abstraction over a MongoDB client. Developers can point resolvers to any datasource.
```

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+    title,
+    description,
+    address,
+    id
+  }
+}
+
+mutation createTask {
+  createTask(input: {title: "complete the walkthrough", description: "complete the GraphQL walkthrough", address: "NA"}) {
+    title,
```
Comment (craicoverflow) [inline walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc]: I don't think it will work with commas?

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc (hunk 9 above)
Comment (wtrocki) [reply walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc:101]: Yes. This is basing on the master branch. We have renamed that to delta to avoid confusion, but I will be able to make this change only after we release 0.14.0 Graphback

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc (hunk 1 above)
Comment (wtrocki) [reply walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc]: I think I will remove this as we not using that anyway

**Comment by wtrocki** [comment]
@mmusil Would you mind facilitating this update with QE team. 
I want QE to be aware that we changing the WT now.

**Review by martinhesko — APPROVED** [review]
LGTM

File: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+. Execute the `start server` command. This command starts the GraphQL server which also serves the client application for simplicity.
+. The application should be opened in a preview window after build is finished.
+
+[type=verification]
+****
+. Check if the website was loaded properly
+. Select the + icon to create a new item
+. On the new screen enter a `name` and `description` and create the task.
+. New task should appear in the task list.
+----
```
Comment (mmusil) [inline walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc]: Is there supposed to be any code snippet? It creates this empty code snipped.
image
```suggestion
```

Comment (wtrocki) [reply walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc]: I have noticed it. It was working in old version of the solution explorer so it might that format changed

**Comment by wtrocki** [comment]
Verification done. Are we ok to integrate this solution pattern for the next release?

**Comment by wtrocki** [comment]
Needs approaval from QE (It will require using code ready for testing)
Needs approval from a lead (CC @tiffanynolan)

Approval from the team already there.

**Comment by wtrocki** [comment]
Screenshot 2020-06-16 at 15 36 34

**Review by craicoverflow — APPROVED** [review]

**Comment by wtrocki** [comment]
Thank you @craicoverflow for review.

**Comment by tiffanynolan** [comment]
@wtrocki I had trouble loading the new SP into a cluster so unfortunately, I'm not able to verify this one for the release. Given the file was replaced, it is also very difficult to be able to see/know what changed. How important is it that this version goes live with the 2.4 release? Given that CodeReady is in this updated version, I'm not confident about adding this one in just yet.

**Comment by wtrocki** [comment]
> @wtrocki I had trouble loading the new SP into a cluster so unfortunately,

No problem. I should have added verification steps.
I used the following instruction
1. In your cluster go to settings`https://solution-explorer.cluster/settings`
2. Put the following URL into the 
https://github.com/integr8ly/solution-patterns.git#update-data-sync-wt
It should look as follows:
Screenshot 2020-06-17 at 09 10 19

> Given the file was replaced, it is also very difficult to be able to see/know what changed. 

This is a completely new walkthrough covering the new generation of the data sync that we used in the demos for the red hat summit and various customers over the last 5 months. We were maintaining this walkthrough separately in our repository.
The previous walkthrough wasn't changed for the last year and it is covering the previous generation of the DataSync

> How important is it that this version goes live with the 2.4 release?

The previous walkthrough will stop working with the latest SSO 7.4 if that is going to be included into 2.4 release. This backport will not be easy as it will broke 1.x RHMI that will not have Keycloak updated so we will need to spent roughly the entire sprint to cover this work.

We currently maintaining both walkthroughs despite old one is not really reflecting what we currently have and it has very challenging backports due to our codebase moved so much over the last year. 
This is taking so much time from the team and also limiting exposure of the new stuff to the users.   

> Given that CodeReady is in this updated version, I'm not confident about adding this one in just yet.

Would removing CodeReady change your opinion?
Little bit more context here:

We have been chatting with various folks and getting feedback about our walkthrough. Main point was that there was actually no interaction with the code in datasync walkthrough and our new platform is really all about people interacting with their editor. I'm aware that CodeReady might bring some `instability` issues as I have experienced that a lot but I hope that thanks to this walkthrough.

We added CodeReady to support training use cases - walkthrough can be used by @evanshortiss to deliver some nice looking mobile-oriented training. 

-----
TL;DR?
We have the following options for 2.4:

1) Remove keycloak setup from the current walkthrough to make it work.
2) Work on accepting this walkthrough that is proven to work with this keycloak version. 

Considering your feedback I think we might actually start with 1 to be secure for the 2.4 release and work with yourself and QE to see what version of the walkthrough they will love to see - we might need to sacrifice `training` path if needed.

**Comment by wtrocki** [comment]
I have added all QE engineers I knew used to test our walkthough to reviewers. I will look for feedback on this before we all succumb to CodeReady based WT.

**Comment by mmusil** [comment]
Actually @wtrocki, these steps with solution explorer settings didn't work when I was creating a cluster for you yesterday, because of a conflict with already present Solution Patterns from different branch used by RHMI master. What I had to do, was get to the `redhat-rhmi-solution-explorer` namespace, Replication Controllers -> pick the RC that's used, change the `WALKTHROUGH_LOCATIONS` environment variable to use a different branch, delete running pod (so the new config is applied) and then the new WT was loaded.

**Comment by wtrocki** [comment]
Yes. I should have used my own fork to simplify testing.

**Comment by wtrocki** [comment]
Update for the progress:

1) We need to have some QE acceptance on this.
@mmusil Would you be able to mention on the next QE meeting. 
2) After QE acceptance we will work with content team to align with the standards that other solutions have. 
3) Our team will do another run thru the walkthrough. 
4) We merge this to the next release

**Comment by wtrocki** [comment]
Update

@mmusil Reported minor label changes in the current version of solution explorer. 
We are working on investigating codeready issues. It looks like clusters with codeready failing to start with any project (not even datasync)

**Comment by mmusil** [comment]
Currently, I'm experiencing some problems with CodeReady. Last week I've updated devfile.yaml in Data Sync Starter repo to work with CRW 2.1.1, this week for some reason those changes don't work and I haven't figured out yet what is the reason for that. I'm having real concerns about using CRW for the Walkthrough considering such instability of CRW.

**Comment by wtrocki** [comment]
@mmusil Thank you for raising concerns. I will get a hold of the RHMI environment and do testing ourselves. I do not think that issue is in the dev file because I have noticed that regular stacks also failing to deploy - always with the source code checkout. 

When developing this walkthrough I have been using OpenShift dedicated with CodeReady installed from their CLI and this was very stable for us - same for keycloak. I think that the ultimate problem is just cluster size/number of services running inside the cluster that affect the overall stability of it. That is why the issues we see are mostly random. 

I will look if we can tie down and confirm that this is only related to the resource issues by getting hands on more powerful cluster/provisioning only CR and Keycloak to it.

**Comment by wtrocki** [comment]
@mmusil Can you point out minor label fixes that were reported last time?

**Comment by wtrocki** [comment]
Update from the extensive amount of investigations and discussions that happened over the last weeks. This WT uses CodeReady. We are aware that CodeReady is not a solidified RHMI service and there are some issues that we are facing when using it with this particular WT. When using this walkthrough CodeReady workspace runs out of the disc space. Default 1GB quota is too small for the amount of development, packages and builds that we need to execute here. 

We were exploring alternatives:

1) Using an image with a prebuild datasync-starter where developers can supply some of the overrides by mounting config maps. This approach will require minor changes in the datasync-starter.
Developers of the walkthrough could use separate (very small repository) that will contain only override files and it will not require packages to be installed. This requires full rebuild of the walkthrough and some effort to create those artifacts.

The small repository will contain:

- GraphQL Model file declaring all the database objects and API that will be build dynamically.
- Authz file declaring authorization rules.
- Keycloak config files that should be injected
- .env file containing environment variables overrides that are not infrastructure (like topics etc.) 
- The template that can be used to deploy the project using **existing** image with appropriate config maps. 
- Entire client-side project that utilizes 
- Some generated client side queries and full graphql schema

DataSync as service (1)



2) Utilizing CodeReady as it is but restricting number of packages we use to not run out of space - this was proven to be difficult. After investigation, we are sure that this is not reliable way to go forward.

3) Utilizing CodeReady but changing 1GB limit to some larger (this is happening during installation phase and we need to increase quota for PV. While this sounds simple it might become challenging on environments with CodeReady installed and it will require some investigation first.

4) Use local workflow when developers need to have Node.js installed locally and build project on their own laptop. This has lots of benefits as we can show much more and utilize user `oc` cmd line that is connected to openshift cluster to execute various operations and show automatic integration with RHMI. 

We already have done this version of the walkthrough, although we will need to review it again and probably improve
https://raw.githubusercontent.com/aerogear/datasync-starter/043c6a9dfd9170e363ed833037bc4156b47c27e9/walkthroughs/2-exploring-datasync-local/disabled_walkthrough.adoc

**Comment by wtrocki** [comment]
We have reworked this walkthrough after feedback that we do not provide much access to the code and the ability to get hands-on the code and features. I believe that option nr 1 could be considered an DataSync "service" that is self-managed and deployed with supplied user configuration.

Looking for feedback on approach 1 before we invest time on implementing it.
Approach nr1 resembles what "other" realtime providers offer so it makes sense to pursue it from the upstream point of view.

**Comment by wtrocki** [comment]
Update. Due to limitations of CRW we decided to move away from it and have separate PR that has datasync specific steps. 

@mmusil Will link PR once it is created.

**Comment by mmusil** [comment]
PR is ready #41 

@wtrocki @tiffanynolan

**Comment by wtrocki** [comment]
Current plans were ported to datasync trello. 
Once we merge other PR we can close this one as we no loner need this version

**Comment by wtrocki** [comment]
Closing as this is outdated.
